File: app/Sources/LocalCaption/ASR/WhisperEngine.swift
import Foundation
import CoreML
import WhisperKit
import LocalCaptionKit

/// Dual-model WhisperKit engine (SPEC §8, decision B4 settled by on-device benchmark):
/// a fast **interim** model for provisional partials + an accurate **final** model for
/// committed captions. Both load once and stay resident.
///
/// Measured on this Mac (CPU+GPU, 7 s clip): tiny.en ≈ 456 ms, small.en ≈ 2.06 s,
/// large-v3-turbo ≈ 3.53 s — so the default is tiny.en partials + small.en finals; turbo
/// is available as an opt-in accuracy mode via the final-model setting.
final class WhisperEngine {
    private var interim: WhisperKit?
    private var final: WhisperKit?

    let interimName: String       // config-friendly, e.g. "tiny.en"
    let finalName: String         // e.g. "small.en" / "large-v3-turbo"
    let interimVariant: String    // repo folder, e.g. "openai_whisper-tiny.en"
    let finalVariant: String

    private let repo = "argmaxinc/whisperkit-coreml"
    private let sr = 16000.0

    private let finalOptions = DecodingOptions(
        task: .transcribe, language: "en",
        skipSpecialTokens: true, withoutTimestamps: true, windowClipTime: 0)
    private let interimOptions = DecodingOptions(task: .transcribe, language: "en",
        temperatureFallbackCount: 0, sampleLength: 128,
        skipSpecialTokens: true, withoutTimestamps: true, wordTimestamps: true,
        windowClipTime: 0)

    init(interimModel: String, finalModel: String) {
        self.interimName = interimModel
        self.finalName = finalModel
        self.interimVariant = WhisperEngine.variant(for: interimModel)
        self.finalVariant = WhisperEngine.variant(for: finalModel)
    }

    var finalLabel: String { finalName }
    var interimLabel: String { interimName }
    var isLoaded: Bool { interim?.tokenizer != nil && final?.tokenizer != nil }

    /// Map a config-friendly model name to its WhisperKit repo variant folder.
    static func variant(for name: String) -> String {
        switch name {
        case "tiny.en": return "openai_whisper-tiny.en"
        case "base.en": return "openai_whisper-base.en"
        case "small.en": return "openai_whisper-small.en"
        case "large-v3": return "openai_whisper-large-v3"
        case "large-v3-turbo": return "openai_whisper-large-v3-v20240930_turbo"
        case "distil-large-v3": return "distil-whisper_distil-large-v3"
        default: return name.contains("_") ? name : "openai_whisper-\(name)"
        }
    }

    private func localFolder(_ variant: String) -> URL {
        AppPaths.models.appendingPathComponent("models/\(repo)/\(variant)", isDirectory: true)
    }

    private func isDownloaded(_ variant: String) -> Bool {
        let fm = FileManager.default
        let f = localFolder(variant)
        guard fm.fileExists(atPath: f.appendingPathComponent("config.json").path) else { return false }
        for part in ["AudioEncoder.mlmodelc", "MelSpectrogram.mlmodelc", "TextDecoder.mlmodelc"] {
            var isDir: ObjCBool = false
            let ok = fm.fileExists(atPath: f.appendingPathComponent(part).path, isDirectory: &isDir)
            if !ok || !isDir.boolValue { return false }
        }
        return true
    }

    /// Ensure both models are on disk (download only what's missing) and load them resident.
    func prepare(onStatus: @escaping (String) -> Void,
                 onDownload: @escaping (String, Double) -> Void) async throws {
        let interimFolder = try await ensure(interimVariant, name: interimName, onStatus: onStatus, onDownload: onDownload)
        let finalFolder = try await ensure(finalVariant, name: finalName, onStatus: onStatus, onDownload: onDownload)

        onStatus("Loading \(interimName) + \(finalName)…")
        interim = try await loadModel(folder: interimFolder)
        // The two scheduling lanes must never share mutable decoder/cache state,
        // including when the same model is selected for both roles.
        final = try await loadModel(folder: finalFolder)

        guard interim?.tokenizer != nil, final?.tokenizer != nil else {
            throw NSError(domain: "WhisperEngine", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "Tokenizer failed to load"])
        }
        // Loading weights does not warm CoreML's first prediction. Do this before
        // capture so first-use compilation cannot consume the live decode budget.
        onStatus("Warming speech models…")
        let silence = [Float](repeating: 0, count: 32000)
        var warmFinal = finalOptions
        warmFinal.temperatureFallbackCount = 0
        for (model, options) in [(interim, interimOptions), (final, warmFinal)] {
            let outcome = await run(model, silence, options: options, budget: nil)
            if case .failure(let message) = outcome {
                throw NSError(domain: "WhisperEngine", code: 2,
                              userInfo: [NSLocalizedDescriptionKey: message])
            }
        }
    }

    private func ensure(_ variant: String, name: String,
                        onStatus: @escaping (String) -> Void,
                        onDownload: @escaping (String, Double) -> Void) async throws -> URL {
        if isDownloaded(variant) { return localFolder(variant) }
        onStatus("Downloading \(name)…")
        return try await WhisperKit.download(variant: variant, downloadBase: AppPaths.models, from: repo) { p in
            onDownload(name, p.fractionCompleted)
        }
    }

    private func loadModel(folder: URL) async throws -> WhisperKit {
        let compute = ModelComputeOptions(
            melCompute: .cpuAndGPU, audioEncoderCompute: .cpuAndGPU,
            textDecoderCompute: .cpuAndGPU, prefillCompute: .cpuAndGPU)
        let wk = try await WhisperKit(WhisperKitConfig(
            modelFolder: folder.path, computeOptions: compute,
            verbose: false, logLevel: .info, prewarm: false, load: false, download: false))
        try await wk.loadModels()
        return wk
    }

    func transcribeInterim(_ audio: [Float]) async -> SpeechOutcome {
        await run(interim, audio, options: interimOptions, budget: 2)
    }

    func transcribeFinal(_ audio: [Float]) async -> SpeechOutcome {
        await run(final, audio, options: finalOptions, budget: nil)
    }

    /// Final accuracy options/filters are retained; windowClipTime is zero because
    /// its default skips <=1 s utterances entirely. Interim decoding has a token
    /// ceiling and cooperative deadline; an outstanding CoreML call is never overlapped
    /// with a replacement decode on the same instance.
    private func run(_ wk: WhisperKit?, _ audio: [Float], options: DecodingOptions,
                     budget: Double?) async -> SpeechOutcome {
        guard let wk else { return .failure("Speech model is not loaded") }
        guard Double(audio.count) / sr > 0.2 else { return .empty }
        let deadline = budget.map { ProcessInfo.processInfo.systemUptime + $0 }
        do {
            try Task.checkCancellation()
            let results: [TranscriptionResult] = try await wk.transcribe(audioArray: audio, decodeOptions: options) { _ in
                if let deadline, ProcessInfo.processInfo.systemUptime >= deadline { return false }
                return nil
            }
            try Task.checkCancellation()
            if let deadline, ProcessInfo.processInfo.systemUptime >= deadline { return .timedOut }
            var kept: [String] = []
            var words: [CaptionWord] = []
            var rejected = false
            var fallbacks = 0
            for result in results {
                fallbacks += Int(result.timings.totalDecodingFallbacks)
                for seg in result.segments {
                    if Filters.isLowQuality(avgLogprob: Double(seg.avgLogprob),
                                            noSpeechProb: Double(seg.noSpeechProb),
                                            compressionRatio: Double(seg.compressionRatio)) {
                        rejected = true
                        continue
                    }
                    let cleaned = Filters.clean(seg.text)
                    if !cleaned.isEmpty {
                        kept.append(cleaned)
                        words += (seg.words ?? []).map {
                            CaptionWord(Filters.clean($0.word), start: Double($0.start), end: Double($0.end))
                        }
                    }
                }
            }
            let text = Filters.clean(kept.joined(separator: " "))
            if text.isEmpty { return rejected ? .filtered : .empty }
            if Filters.isHallucination(text) { return .filtered }
            return .success(text: text, words: words, fallbacks: fallbacks)
        } catch is CancellationError { return .cancelled }
        catch { return .failure(error.localizedDescription) }
    }
}

File: windows/src/LocalCaption.Asr/WhisperEngine.cs
using LocalCaption.Core.Audio;
using LocalCaption.Core.Captions;
using Whisper.net;
using Whisper.net.LibraryLoader;

namespace LocalCaption.Asr;

/// <summary>What a loaded engine ended up using, for the log and the Settings readout.</summary>
/// <param name="Backend">The backend that was <i>asked for</i> after probing.</param>
/// <param name="Library">
/// The native library whisper.cpp <i>actually</i> loaded. Not the same thing: Whisper.net
/// walks its runtime order and falls back to the CPU library without raising anything when
/// a GPU one cannot load — a missing CUDA dependency looks exactly like a slow machine.
/// §5.2 requires the active backend to be visible rather than merely felt, and this is the
/// field that makes it so.
/// </param>
public sealed record EngineInfo(
    string InterimModel, string FinalModel, AsrBackend Backend,
    bool WordTimings, int Threads, string RuntimeInfo, string Library = "unknown")
{
    /// <summary>
    /// True when CUDA was asked for and something else loaded — the §5.8 symptom that would
    /// otherwise present only as captions quietly getting slower.
    /// </summary>
    /// <remarks>
    /// CUDA only. Metal is not a separate runtime library in Whisper.net — the macOS build
    /// reports it in the <c>Cpu</c> slot — so asking the same question there would raise a
    /// false alarm on every Mac development run.
    /// </remarks>
    public bool FellBack =>
        Backend is AsrBackend.Cuda &&
        !Library.Equals(nameof(AsrBackend.Cuda), StringComparison.OrdinalIgnoreCase);

    public override string ToString() =>
        $"{Backend.ToString().ToLowerInvariant()}→{Library.ToLowerInvariant()} · interim={InterimModel} final={FinalModel} · " +
        $"threads={Threads} · word-timings={(WordTimings ? "dtw" : "none")}" +
        (FellBack ? "  ⚠ fell back to the CPU library" : "");
}

/// <summary>
/// Dual-model whisper.cpp engine: a fast <b>interim</b> model for provisional partials and
/// an accurate <b>final</b> model for committed captions. Both load once and stay resident.
/// Replaces the macOS <c>WhisperEngine.swift</c>.
/// </summary>
/// <remarks>
/// <para><b>The two lanes never share decoder state.</b> <c>whisper_context</c> is not
/// thread-safe, so each lane gets its own factory <i>and</i> its own processor — including
/// when the same model name is chosen for both roles, which is the case that looks safe to
/// share and is not. This is the rule that came out of the macOS <c>e5ec3bd</c> fix
/// (SPEC-WINDOWS.md §5.1), and it is why <see cref="Prepare"/> builds two of everything
/// rather than reusing one factory.</para>
/// <para><b>DTW is a factory-level option, not a per-call one.</b> Word timings come from
/// <see cref="WhisperFactoryOptions.UseDtwTimeStamps"/> plus the model's alignment-head
/// preset, so they must be decided before the weights load. The interim lane needs them —
/// <see cref="RollingCaption"/> merges sliding windows purely by word midpoint — while the
/// final lane does not, and pays nothing for skipping them.</para>
/// </remarks>
public sealed class WhisperEngine : IAsyncDisposable
{
    private const double SampleRate = 16000.0;
    /// <summary>Below this, a window is not worth a decode — it is a click, not a word.</summary>
    private const double MinimumSeconds = 0.2;

    private WhisperFactory? _interimFactory;
    private WhisperFactory? _finalFactory;
    private WhisperProcessor? _interim;
    private WhisperProcessor? _final;

    /// <summary>The longest prompt worth sending: whisper.cpp keeps about 224 tokens of it.</summary>
    private const int PromptLimit = 600;

    public string InterimName { get; }
    public string FinalName { get; }

    /// <summary>The initial prompt built from <c>asr.vocabulary</c>, or null when there is none.</summary>
    public string? Prompt { get; }

    /// <summary>Beam width on the final lane; below 2 means greedy.</summary>
    public int FinalBeamSize { get; }
    public AsrBackend RequestedBackend { get; }
    public int Threads { get; }
    public EngineInfo? Info { get; private set; }
    public bool IsLoaded => _interim is not null && _final is not null;

    public WhisperEngine(string interimModel, string finalModel,
                         AsrBackend backend = AsrBackend.Auto, int threads = 0, string? vocabulary = null,
                         int finalBeamSize = 0)
    {
        InterimName = interimModel;
        FinalName = finalModel;
        Prompt = BuildPrompt(vocabulary);
        FinalBeamSize = Math.Clamp(finalBeamSize, 0, 8);
        RequestedBackend = backend;
        // 0 means "physical cores" (§9.2). Environment.ProcessorCount counts logical
        // processors, and oversubscribing whisper.cpp with SMT siblings costs throughput.
        Threads = threads > 0 ? threads : Math.Max(1, Environment.ProcessorCount / 2);
    }

    /// <summary>
    /// Download whatever is missing, load both models resident, and warm them.
    /// </summary>
    public async Task<EngineInfo> PrepareAsync(string modelsDirectory,
                                               Action<string>? onStatus = null,
                                               ModelDownloader.Progress? onDownload = null,
                                               CancellationToken cancellationToken = default)
    {
        var interimSpec = ModelCatalog.Resolve(InterimName);
        var finalSpec = ModelCatalog.Resolve(FinalName);

        using (var downloader = new ModelDownloader())
        {
            foreach (var spec in Distinct(interimSpec, finalSpec))
            {
                var path = ModelCatalog.PathFor(spec, modelsDirectory);
                if (!File.Exists(path)) onStatus?.Invoke($"Downloading {spec.Name}…");
                await downloader.EnsureAsync(spec, modelsDirectory, onDownload, cancellationToken)
                                .ConfigureAwait(false);
            }
        }

        var backend = BackendProbe.Resolve(RequestedBackend);
        onStatus?.Invoke($"Loading {InterimName} + {FinalName}…");

        // The interim lane needs word timings; the final lane does not and skips the cost.
        _interimFactory = CreateFactory(ModelCatalog.PathFor(interimSpec, modelsDirectory),
                                        backend, interimSpec.Heads, wordTimings: true);
        _finalFactory = CreateFactory(ModelCatalog.PathFor(finalSpec, modelsDirectory),
                                      backend, finalSpec.Heads, wordTimings: false);

        _interim = BuildInterim(_interimFactory);
        _final = BuildFinal(_finalFactory);

        // Loading weights does not warm the first prediction — whisper.cpp still has to
        // compile its GPU kernels. Doing that here rather than on first speech is the
        // difference between a clean first caption and a multi-second stall.
        onStatus?.Invoke("Warming speech models…");
        var silence = new float[(int)(2 * SampleRate)];
        foreach (var processor in new[] { _interim, _final })
        {
            await foreach (var _ in processor.ProcessAsync(silence, cancellationToken).ConfigureAwait(false)) { }
        }

        Info = new EngineInfo(InterimName, FinalName, backend,
                              WordTimings: interimSpec.Heads != WhisperAlignmentHeadsPreset.None,
                              Threads, WhisperFactory.GetRuntimeInfo() ?? "unknown",
                              Library: LoadedLibrary());
        onStatus?.Invoke($"Ready — {Info}");
        return Info;
    }

    /// <summary>
    /// Turn a comma-separated list into the sentence whisper.cpp is primed with.
    /// </summary>
    /// <remarks>
    /// The decoder treats its prompt as "what was said just before", so a name that appears
    /// there is a name it will prefer to spell that way again. It is a nudge, not a
    /// dictionary: it fixes "Chukwu Emeka" → "Chukwuemeka", it does not teach a language the
    /// model was never trained on.
    /// </remarks>
    public static string? BuildPrompt(string? vocabulary)
    {
        if (string.IsNullOrWhiteSpace(vocabulary)) return null;

        var terms = vocabulary.Split([',', ';', '\n', '\r'], StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
                              .Where(t => t.Length > 0).Distinct(StringComparer.OrdinalIgnoreCase);
        var list = string.Join(", ", terms);
        if (list.Length == 0) return null;
        if (list.Length > PromptLimit) list = list[..PromptLimit];
        return $"{PromptLead}: {list}.";
    }

    /// <summary>
    /// Which native library Whisper.net settled on, read after the first factory is built.
    /// </summary>
    /// <remarks>
    /// Whisper.net resolves this once per process, on first load, by trying each library in
    /// its runtime order and moving on when one will not load. Nothing is thrown and nothing
    /// is logged, so without reading it back a CUDA build that silently ran on the CPU is
    /// indistinguishable from a CUDA build that was simply slow (§5.2).
    /// </remarks>
    private static string LoadedLibrary()
    {
        try { return RuntimeOptions.LoadedLibrary?.ToString() ?? "cpu"; }
        catch (Exception) { return "unknown"; }
    }

    /// <summary>Two specs, or one when both roles chose the same model.</summary>
    private static IEnumerable<ModelSpec> Distinct(ModelSpec a, ModelSpec b) =>
        a.FileName == b.FileName ? [a] : [a, b];

    private WhisperFactory CreateFactory(string modelPath, AsrBackend backend,
                                         WhisperAlignmentHeadsPreset heads, bool wordTimings)
    {
        if (!File.Exists(modelPath))
            throw new FileNotFoundException($"Speech model weights are missing: {modelPath}", modelPath);

        var options = WhisperFactoryOptions.Default;
        options.UseGpu = BackendProbe.UsesGpu(backend);
        options.UseDtwTimeStamps = wordTimings && heads != WhisperAlignmentHeadsPreset.None;
        options.HeadsPreset = options.UseDtwTimeStamps ? heads : WhisperAlignmentHeadsPreset.None;
        return WhisperFactory.FromPath(modelPath, options);
    }

    /// <summary>
    /// Interim decoding: greedy, no context carried between windows, token timestamps on.
    /// Speed is the whole point — an interim result that misses its budget is discarded.
    /// </summary>
    private WhisperProcessor BuildInterim(WhisperFactory factory) => factory.CreateBuilder()
        .WithLanguage("en")
        .WithThreads(Threads)
        .WithTokenTimestamps()
        .WithProbabilities()
        // Each interim window is an independent 6 s tail, so carrying context from the last
        // one would let a hallucination propagate forward across every later window.
        .WithNoContext()
        .WithGreedySamplingStrategy(g => g.WithBestOf(1))
        .WithTemperature(0f)
        .Build();

    /// <summary>
    /// Final decoding: accuracy over latency. This output is what gets written to the
    /// transcript and can never be revised, so it keeps temperature fallback enabled.
    /// </summary>
    /// <remarks>
    /// The vocabulary prompt goes here and only here. The interim lane decodes overlapping
    /// six-second windows several times a second, where a prompt is both wasted time and one
    /// more thing for a fragment of silence to be "completed" into.
    /// </remarks>
    private WhisperProcessor BuildFinal(WhisperFactory factory)
    {
        var builder = factory.CreateBuilder()
            .WithLanguage("en")
            .WithThreads(Threads)
            .WithProbabilities()
            .WithNoContext();
        if (Prompt is not null) builder = builder.WithPrompt(Prompt);

        // Final lane only, for the same reason as the prompt: an interim result is thrown
        // away half a second later, so spending twice the decode on it buys nothing.
        if (FinalBeamSize > 1) builder = builder.WithBeamSearchSamplingStrategy(beam => beam.WithBeamSize(FinalBeamSize));
        return builder.Build();
    }

    // ── Decoding ─────────────────────────────────────────────────────────────────────────

    public Task<SpeechOutcome> TranscribeInterimAsync(IReadOnlyList<float> audio, CancellationToken cancellationToken = default) =>
        RunAsync(_interim, audio, wantWords: true, cancellationToken);

    public async Task<SpeechOutcome> TranscribeFinalAsync(IReadOnlyList<float> audio, CancellationToken cancellationToken = default)
    {
        var outcome = await RunAsync(_final, audio, wantWords: false, cancellationToken).ConfigureAwait(false);

        // A primed decoder handed near-silence will sometimes recite its prompt. That is a
        // hallucination with a known text, so it can be caught exactly.
        return outcome is SpeechOutcome.Success success && EchoesPrompt(success.Text)
            ? new SpeechOutcome.Filtered()
            : outcome;
    }

    private bool EchoesPrompt(string text)
    {
        if (Prompt is null) return false;
        var said = text.Trim().TrimEnd('.', ' ');

        // Narrow on purpose. Someone may genuinely say two listed names in a row, and losing
        // real speech is the worse error — so only the prompt's own framing, or most of the
        // prompt recited whole, counts as an echo.
        return said.StartsWith(PromptLead, StringComparison.OrdinalIgnoreCase) ||
               (said.Length >= Prompt.Length * 0.6 && Prompt.Contains(said, StringComparison.OrdinalIgnoreCase));
    }

    private const string PromptLead = "Names and terms";

    private static async Task<SpeechOutcome> RunAsync(WhisperProcessor? processor, IReadOnlyList<float> audio,
                                                      bool wantWords, CancellationToken cancellationToken)
    {
        if (processor is null) return new SpeechOutcome.Failure("Speech model is not loaded");
        if (audio.Count / SampleRate <= MinimumSeconds) return new SpeechOutcome.Empty();

        var samples = audio as float[] ?? [.. audio];
        var kept = new List<string>();
        var words = new List<CaptionWord>();
        var rejected = false;

        try
        {
            await foreach (var segment in processor.ProcessAsync(samples, cancellationToken).ConfigureAwait(false))
            {
                // §5.6: whisper.cpp reports no_speech_prob, but avg_logprob and
                // compression_ratio have to be derived before the gate can run.
                var logprob = SegmentQuality.AverageLogprob(
                    (segment.Tokens ?? []).Select(t => (double)t.ProbabilityLog));
                var compression = SegmentQuality.CompressionRatio(segment.Text ?? "");

                if (Filters.IsLowQuality(logprob, segment.NoSpeechProbability, compression))
                {
                    rejected = true;
                    continue;
                }

                var cleaned = Filters.Clean(segment.Text ?? "");
                if (cleaned.Length == 0) continue;

                kept.Add(cleaned);
                if (wantWords) words.AddRange(WordTimings.FromTokens(segment));
            }
        }
        catch (OperationCanceledException) { return new SpeechOutcome.Cancelled(); }
        catch (Exception e) { return new SpeechOutcome.Failure(e.Message); }

        var text = Filters.Clean(string.Join(" ", kept));
        if (text.Length == 0) return rejected ? new SpeechOutcome.Filtered() : new SpeechOutcome.Empty();
        if (Filters.IsHallucination(text)) return new SpeechOutcome.Filtered();

        // whisper.cpp has no temperature-fallback counter to report, unlike WhisperKit.
        return new SpeechOutcome.Success(text, words, Fallbacks: 0);
    }

    public async ValueTask DisposeAsync()
    {
        if (_interim is not null) await _interim.DisposeAsync().ConfigureAwait(false);
        if (_final is not null) await _final.DisposeAsync().ConfigureAwait(false);
        _interimFactory?.Dispose();
        _finalFactory?.Dispose();
        _interim = null;
        _final = null;
        _interimFactory = null;
        _finalFactory = null;
    }
}
